### PR TITLE
Qa mars

### PR DESCRIPTION
**Pull request by isu2**
Add Excel file for review

### CODE REVIEW

File: Features/ProfilePage.feature [hunk 1]
```diff
@@ -0,0 +1,14 @@
+Feature: Profile Management
+
+Background:
+    Given I am on the home page
```
Comment (amanbansalnz) [inline Features/ProfilePage.feature]: you should split the features into two feature files one for language and one for skills!

File: Features/ProfilePage.feature [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    When I click on the sign-in button
+    And I enter the username "user@example.com" and password "password123"
```
Comment (amanbansalnz) [inline Features/ProfilePage.feature]: Please increase your coverage around the features you  are testing here is an idea of what is missing 

• Negative testing with valid input
• Negative testing with invalid input
• Destructive testing

- E.g of above scenarios
- Happy path tests check basic functionality and the acceptance criteria of the API. We later extend positive tests to include optional parameters and extra functionality.
- 
- The next group of tests is negative testing where we expect the application to gracefully handle problem scenarios with both valid user input (for example, trying to add an existing username) and invalid user input (trying to add a username which is null).
- 
- Destructive testing is a deeper form of negative testing where we intentionally attempt to break the API to check its robustness (for example, sending a huge payload body in an attempt to overflow the system).
- 
- If some scenarios does not make sense to add you can you should know why it does not make  sense to add.

File: Features/ProfilePage.feature [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    Then I click the login button
+
+Scenario: Manage Profile
```
Comment (amanbansalnz) [inline Features/ProfilePage.feature]: Please use tables  and pass values to your scenarios for example look at the below scenario 
This way you will get exposure to use them  and will know how to pass test data to the tests
You all need to pass the values down from feature files to the steps see examples below


Example 

	Given I logged into the portal
	When I click on tab Education
	And I add my education 
	Then I am able to see my education'

Would  change to 


@Education
Scenario Outline: 1-Add education
	Given I logged into the portal
	When I click on tab Education
	And I add my education including '<Country>', '<University>', '<Title>', '<Degree>', '<Graduation Year>'
	Then I am able to see my education details including '<Country>', '<University>', '<Title>', '<Degree>', '<Graduation Year>'

	Examples:
	| Country     | University                      | Title  | Degree  | Graduation Year |
	| New Zealand | Whitireia Community Polytechnic | M.Tech | Diploma | 2021            |
	| Venezuela   | Hhutech University              | B.A    | Diploma | 2011            |

See  https://cucumber.io/docs/gherkin/reference/

File: Pages/ProfilePage.cs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+using SeleniumExtras.WaitHelpers;
+
+namespace QA_Mars_OnboardingTaskSpecflow.Pages
+{
+    public class ProfilePage
+    {
+        private readonly IWebDriver driver;
+        private WebDriverWait wait;
+        private readonly By btnAddNewLangauageLocator = By.XPath("//*[@id=\"account-profile-section\"]/div/section[2]/div/div/div/div[3]/form/div[2]/div/div[2]/div/table/thead/tr/th[3]/div");
+        private readonly By txaLanguageLocator = By.XPath("//*[@id=\"account-profile-section\"]/div/section[2]/div/div/div/div[3]/form/div[2]/div/div[2]/div/div/div[1]/input");
```
Comment (amanbansalnz) [inline Pages/ProfilePage.cs]: please change all variables to return IWebElement rather than BY for example  By txaLanguageLocator = By.XPath("//*[@id=\"account-profile-section\"]/div/section[2]/div/div/div/div[3]/form/div[2]/div/div[2]/div/table/thead/tr/th[3]/div")

to 

private IWebElement txaLanguageLocator = driver.FindElement( By.XPath("//*[@id=\"account-profile-section\"]/div/section[2]/div/div/div/div[3]/form/div[2]/div/div[2]/div/table/thead/tr/th[3]/div")));

Comment (amanbansalnz) [reply Pages/ProfilePage.cs]: you can see 
https://docs.specflow.org/projects/specflow/en/latest/ui-automation/Selenium-with-Page-Object-Pattern.html 
 for example

Comment (amanbansalnz) [reply Pages/ProfilePage.cs]: please fix all pages

File: Steps/ProfilePageStepDefinitions.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        [When(@"I edit the language ""(.*)"" to ""(.*)""")]
+        public void WhenIEditTheLanguageTo(string fromLanguage, string toLanguage)
+        {
+            profilePage.EditLanguage(fromLanguage, toLanguage);
+        }
+
+        // Corrected method name and signature; assumes no parameter needed
+        [Given(@"I click on the ""Delete Language"" button")]
+        public void GivenIClickOnDeleteLanguage()
+        {
```
Comment (amanbansalnz) [inline Steps/ProfilePageStepDefinitions.cs:170]: you are missing your asserts, usually you would add this in the then! please add assets

File: Steps/ProfilePageStepDefinitions.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        private readonly IWebDriver driver;
+        private readonly ProfilePage profilePage;
+
+        public ProfilePageStepDefinitions(IWebDriver driver)
+        {
+            this.driver = driver;
+            profilePage = new ProfilePage(driver);
+        }
+
```
Comment (amanbansalnz) [inline Steps/ProfilePageStepDefinitions.cs:19]: you also need to think about state State

example before adding language you need to make sure there a no existing languages as this could result in false positive test pass
you will need to think about this for all the scenarios

the state is say that you have a test that is adding a language english, the second time you run this test you should delete any existing language so that the state is reset, this will guarantee that your test does not produce false negatives and passes always
you can assume that test execution order is not always guaranteed
imagine if you run the tests  in parallel mode the all tests will run at the same time
so you need to also make sure you do not wipe out other tests data
hence why when you remove data the best thing to do is delete only data you added for the test to be correct
e.g test 1 added english and french as language test 2 added hindi then when you reset data you already know what language you are adding for each test
there for on reset of the two test you should delete english and french for test 1 and hindi for test 2

So to sum up you should delete all data if there is any before the start of any test and then after each test you should delete the data that a particular test is adding!
This is ensure that upon staring the test pack you have reset state and upon finishing each test deletes their data that they added!

**Review by amanbansalnz — CHANGES_REQUESTED** [review]
please fix the suggested changes

File: Features/ProfilePage.feature (hunk 1 above)
Comment (isu2) [reply Features/ProfilePage.feature]: Thank you for your feedback! This has been fixed.

File: Features/ProfilePage.feature (hunk 2 above)
Comment (isu2) [reply Features/ProfilePage.feature]: Thank you for your feedback! This has been fixed.

File: Features/ProfilePage.feature (hunk 3 above)
Comment (isu2) [reply Features/ProfilePage.feature]: Thank you for your feedback! This has been fixed.

File: Steps/ProfilePageStepDefinitions.cs (hunk 5 above)
Comment (isu2) [reply Steps/ProfilePageStepDefinitions.cs:170]: Thank you for your feedback! This has been fixed.

File: Steps/ProfilePageStepDefinitions.cs (hunk 6 above)
Comment (isu2) [reply Steps/ProfilePageStepDefinitions.cs:19]: Thank you for your feedback! This has been fixed.

Comment (amanbansalnz) [reply Steps/ProfilePageStepDefinitions.cs:19]: can you explain how did you implement this please

Comment (isu2) [reply Steps/ProfilePageStepDefinitions.cs:19]: Hi Aman,
Thank you for reviewing the task. According to your this review feedback, I incorporated and automated another test scenario to identify existing/duplicate records and deleted records via a separate scenario. If I misunderstood anything, please let me know. I highly appreciate your support. Thank you! ☺️

File: Pages/ProfilePage.cs [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+       IWebElement btnUpdateLanaguage = driver.FindElement(btnUpdateLanaguageLocator);
+      btnUpdateLanaguage.Click();
+
+       }
+
+       catch(WebDriverTimeoutException ex){
+            Console.WriteLine("Timeout waiting for element: " + ex.Message);
+       }
+       //wait.Until(ExpectedConditions.ElementIsVisible(txaEditLanguageLocator)); 
+
```
Comment (amanbansalnz) [inline Pages/ProfilePage.cs:144]: make sure you review all commented out code for further ref

**Review by amanbansalnz — APPROVED** [review]
looks good make sure you review all commented out code for next task
